common/autoware_universe_utils/src/geometry/random_concave_polygon.cpp

```diff
@@ -0,0 +1,524 @@
+  eBreaking.valid = false;
+
+  for (const auto & edge : P_with_edges.edges) {
+    if (edge.valid) {
+      Point2d nearest_node = getNearestNode(Q, edge);
+      double distance = dist(edge, nearest_node);
+      if (distance < min_distance) {
+        min_distance = distance;
+        eBreaking = edge;
+      }
+    }
+  }
+
+  return eBreaking;
+}
+
+bool isValidInsertion(const LinearRing2d & new_ring)
+{
+  for (size_t j = 0; j < new_ring.size(); ++j) {
+    const Point2d & a = new_ring[j];
+    const Point2d & b = new_ring[(j + 1) % new_ring.size()];
+    const Point2d & c = new_ring[(j + 2) % new_ring.size()];
+
+    double cross_product = (b.x() - a.x()) * (c.y() - b.y()) - (b.y() - a.y()) * (c.x() - b.x());
+    if (cross_product <= 0) {
+      return true;
+    }
+  }
+  return false;
+}
+
+void updateEdges(PolygonWithEdges & P_with_edges)
+{
+  auto & ring = P_with_edges.polygon.outer();
+  P_with_edges.edges.clear();  // Clear the old edges
+  for (size_t i = 0; i < ring.size(); ++i) {
+    Edge e = {ring[i], ring[(i + 1) % ring.size()]};
+    P_with_edges.edges.push_back(e);  // Add the new edges
+  }
+}
+
+void insertNode(PolygonWithEdges & P_with_edges, const Point2d & w)
+{
+  auto & ring = P_with_edges.polygon.outer();  // Get the outer ring of the polygon
+
+  // Check if the point `w` already exists in the ring
+  auto it = std::find_if(
+    ring.begin(), ring.end(), [&](const Point2d & p) { return p.x() == w.x() && p.y() == w.y(); });
+
+  if (it != ring.end()) {
+    // Point already exists, so skip insertion
+    return;
+  }
+
+  // If not found, insert the new node `w`
+  for (size_t i = 0; i < ring.size(); ++i) {
+    LinearRing2d new_ring = ring;
+    new_ring.insert(new_ring.begin() + (i + 1), w);  // Insert the new node `w`
+    if (isValidInsertion(new_ring)) {
+      ring = std::move(new_ring);  // Update the polygon's ring with the new ring
+      updateEdges(P_with_edges);   // Update the edges in `P_with_edges`
+    }
+    return;
+  }
+}
+
+void removeNode(std::vector<Point2d> & Q, const Point2d & w)
+{
+  const double epsilon = 1e-9;  // Tolerance for comparing points
+
+  Q.erase(
+    std::remove_if(
+      Q.begin(), Q.end(),
+      [&](const Point2d & p) {
+        return std::abs(p.x() - w.x()) < epsilon && std::abs(p.y() - w.y()) < epsilon;
+      }),
+    Q.end());
+}
+
+void markValidEdges(PolygonWithEdges & P_with_edges, const std::vector<Point2d> & Q)
+{
+  for (auto & edge : P_with_edges.edges) {
+    if (isValid(edge, P_with_edges.polygon, Q)) {
+      edge.valid = true;
+    }
+  }
+}
+
+enum Orientation { COLLINEAR = 0, CLOCKWISE = 1, COUNTERCLOCKWISE = 2 };
+
+Orientation orientation(const Point2d & p, const Point2d & q, const Point2d & r)
+{
+  double val = (q.y() - p.y()) * (r.x() - q.x()) - (q.x() - p.x()) * (r.y() - q.y());
+  if (val == 0) return COLLINEAR;
+  return (val > 0) ? CLOCKWISE : COUNTERCLOCKWISE;
+}
+
+// needed as boost::convex_hull seems not working with defined Polygon2d
```

Comment:
We need to look into this because `boost::geometry::convex_hull()` seems to already be used in many places in autoware. FYI [this test](https://github.com/autowarefoundation/autoware.universe/blob/b90556255030cd4e64fb0a3336152f84011adcab/common/autoware_universe_utils/test/src/geometry/test_geometry.cpp#L2667) passes.

Reply:
Thank you for the review, I have just checked it again, it seems I have to declare Strategy to make it work. I have changed the code so custom convexHull function is not needed anymore, will push it in my next commit. Thanks again!